### PR TITLE
LibGL+LibGPU+LibSoftGPU: Introduce abstract device interface and dynamically load GPU driver

### PR DESCRIPTION
**Pull request by sunverwerth**
Big refactor to cut hard dependencies between LibGL and LibSoftGPU and to make it possible to dynamically load different GPU backends.

TODO:
- [x] Manage .so lifetime. Currently it just gets loaded but never unloaded
- [x] Make the build system always install LibSoftGPU. I hacked a dependency into 3DFileViewer just so it ends up in /us/lib.
- [x] ~See whether it's possible to not have to pledge PROT_EXEC and unveil /usr/lib somehow~ That's just how it is

### CODE REVIEW

File: Userland/Libraries/LibGPU/LightModelParameters.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+#include <LibGPU/Enums.h>
+#include <LibGfx/Vector4.h>
+
+namespace GPU {
+
+struct LightModelParameters {
+    FloatVector4 scene_ambient_color { 0.2f, 0.2f, 0.2f, 1.0f };
+    bool viewer_at_infinity { false };
+    unsigned int single_color { 0x81F9 }; // This is the value of `GL_SINGLE_COLOR`. Considering we definitely don't leak gl.h stuff into here, we fix it to the gl.h macro value.
```
Comment (GMTA) [inline Userland/Libraries/LibGPU/LightModelParameters.h:17]: This was replaced by `color_control` below

File: Userland/Libraries/LibGPU/RasterPosition.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2022, Stephan Unverwerth <s.unverwerth@serenityos.org>
```
Comment (GMTA) [inline Userland/Libraries/LibGPU/RasterPosition.h:2]: Could you readd my copyright line here?

File: Userland/Libraries/CMakeLists.txt [hunk 3]
```diff
@@ -23,6 +23,7 @@ add_subdirectory(LibGemini)
 add_subdirectory(LibGfx)
 add_subdirectory(LibGL)
 add_subdirectory(LibGUI)
+add_subdirectory(LibGPU)
```
Comment (GMTA) [inline Userland/Libraries/CMakeLists.txt:26]: This should be swapped with `LibGUI` to keep an alphabetical order

**Comment by sunverwerth** [comment]
First PR check only fails because of the kernel bug that was fixed today

File: Userland/Libraries/LibGPU/Driver.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    if (it == s_driver_path_map.end())
+        return Error::from_string_literal("The requested GPU driver was not found in the list of allowed driver libraries");
+
+    auto lib = dlopen(it->value.characters(), RTLD_NOW);
+    if (!lib)
+        return Error::from_string_literal("The library for the requested GPU driver could not be opened");
+
+    auto serenity_gpu_create_device = reinterpret_cast<serenity_gpu_create_device_t>(dlsym(lib, "serenity_gpu_create_device"));
+    if (!serenity_gpu_create_device) {
+        dlclose(lib);
```
Comment (creator1creeper1) [inline Userland/Libraries/LibGPU/Driver.cpp:51]: Can this just be wrapped in an RAII class?

File: Userland/Libraries/LibGPU/Image.h [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+    virtual void write_texels(unsigned layer, unsigned level, Vector3<unsigned> const& offset, Vector3<unsigned> const& size, void const* data, ImageDataLayout const& layout) = 0;
+    virtual void read_texels(unsigned layer, unsigned level, Vector3<unsigned> const& offset, Vector3<unsigned> const& size, void* data, ImageDataLayout const& layout) const = 0;
+    virtual void copy_texels(Image const& source, unsigned source_layer, unsigned source_level, Vector3<unsigned> const& source_offset, Vector3<unsigned> const& size, unsigned destination_layer, unsigned destination_level, Vector3<unsigned> const& destination_offset) = 0;
+
+    void const* device_token() const { return m_device_token; }
+    bool has_same_device_token(Image const& other) const { return other.device_token() == device_token(); }
+
+private:
+    void const* const m_device_token { nullptr };
```
[the remark this answers is not in the document]
Comment (sunverwerth) [reply Userland/Libraries/LibGPU/Image.h:32]: Could be. I left these intentionally vague so different implementations can use whatever scheme they like.

File: Userland/Libraries/LibGPU/Driver.cpp (hunk 4 above)
Comment (sunverwerth) [reply Userland/Libraries/LibGPU/Driver.cpp:51]: It is already. If you look a few lines below we inject this into the `Driver` class, but this class can only be instantiated if we have a library that follows our spec.

File: Userland/Libraries/LibGPU/Image.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (creator1creeper1) [reply Userland/Libraries/LibGPU/Image.h:32]: It seems reasonable to me to expect "device tokens" to be `Device *`s

[the remark this answers is not in the document]
Comment (sunverwerth) [reply Userland/Libraries/LibGPU/Image.h]: renamed this to `ownership_token`

**Pull request merged**